### PR TITLE
[SYCL] Guard access to sampler and kernel objects with a mutex

### PR DESCRIPTION
**Pull request by againull**

### CODE REVIEW

**Comment by againull** [comment]
@smaslov-intel Could you please take a look.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 1]
```diff
@@ -1386,6 +1388,12 @@ struct _pi_kernel : _pi_object {
     return true;
   }
 
+  void retain() {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:1391]: Please add a comment explaining that the caller must lock access to the kernel and program.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -4843,47 +4859,55 @@ pi_result piKernelRetain(pi_kernel Kernel) {
...
+  // of.
+  std::scoped_lock Lock(Kernel->Mutex, Kernel->Program->Mutex);
+  Kernel->retain();
   return PI_SUCCESS;
 }
 
 pi_result piKernelRelease(pi_kernel Kernel) {
 
   PI_ASSERT(Kernel, PI_INVALID_KERNEL);
+  auto KernelProgram = Kernel->Program;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4872]: Can you move this under the lock below?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -4901,6 +4925,18 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
...
+  {
+    std::scoped_lock QueueLock(Queue->Mutex);
+    if (auto Res = Queue->Context->getAvailableCommandList(
+            Queue, CommandList, false /* UseCopyEngine */,
+            true /* AllowBatching */))
+      return Res;
+  }
+
+  // Lock automatically releases when this goes out of scope.
+  std::scoped_lock Lock(Queue->Mutex, Kernel->Mutex);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4939]: is it OK that the "Queue" can be operated by another thread after a suitable command-list is found above? why don't we hold the queue lock for the duration of this entire function?

Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4990]: Fixed this. After the change https://github.com/intel/llvm/pull/6168 we can lock queue and kernel simultaneously for the scope of this function.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 1 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.hpp:1399]: Added comment.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Moved under the lock, thx.

**Review by againull — COMMENTED** [review]
Sorry for delayed update. I was able to address your comment after https://github.com/intel/llvm/pull/6168 got landed.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -5074,12 +5098,11 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
...
-  // that the Kernel is in use. Once the event has been signalled, the
-  // code in Event.cleanup() will do a piReleaseKernel to update
-  // the reference count on the kernel, using the kernel saved
-  // in CommandData.
-  PI_CALL(piKernelRetain(Kernel));
+  // Increment the reference count of the Kernel and indicate that the Kernel is
+  // in use. Once the event has been signalled, the code in Event.cleanup() will
+  // do a piReleaseKernel to update the reference count on the kernel, using the
+  // kernel saved in CommandData.
+  Kernel->retain();
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5105]: Does this hold a lock of Kernel's program?

Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5105]: Good catch, fixed, thanks.

**Review by smaslov-intel — APPROVED** [review]

**Pull request merged**